src/search/landmarks/landmark_heuristic.cc

```diff
@@ -152,9 +190,37 @@ void LandmarkHeuristic::generate_preferred_operators(
 }
 
 int LandmarkHeuristic::compute_heuristic(const State &ancestor_state) {
+    /*
+      The path-dependent landmark heuristics are somewhat ill-defined for states
+      not reachable from the initial state of a planning task. We therefore
+      assume here that they are only called for reachable states. Under this
+      view it is correct that the heuristic declares all states as dead-ends if
+      the landmark graph of the initial state has a cycle of natural orderings.
+
+      In the paper on landmark progression (Büchner et al., 2023) we suggest a
+      way to deal with unseen (incl. unreachable) states: These states have zero
+      information, i.e., all landmarks are past and none are future. With this
+      definition, heuristics should yield 0 for all states without expanded
+      paths from the initial state. It would be nice to close the gap between
+      theory and implementation, but we currently don't know how.
+      TODO: Maybe we can find a cleaner way to deal with this once we have a
+       proper understanding of the theory. What component is responsible to
+       detect unsolvability due to cycles of natural orderings or other reasons?
+       How is this signaled to the heuristic(s)? Is adding an unsatisfiable
+       landmark to the landmark graph an option? But this requires changing the
+       landmark graph after construction which we try to avoid at the moment on
+       the implementation side.
+    */
+    if (initial_landmark_graph_has_cycle_of_natural_orderings) {
+        log << "Found a cycle of natural (or stronger) landmark orderings. By "
+            << "the definition of natural orderings, the problem is unsolvable."
+            << endl;
```

Comment:
I am not sure I would want to make this output here. I suggest just returning infinity, but possibly including an output along the lines of this in the code that computes the flag. I find output in a "preprocessing" stage more palatable than "inside" the heuristic computation. In general, at least in future code, I'd also always guard output like this with a verbosity test, but I didn't check how we do it with other landmark-related outputs.

Reply:
Sure, I'll remove the comment in the heuristic. You write "including an output along the lines of this", so is there something in particular you would change about it?

Reply:
It could be briefer, but I don't feel strongly about this. I might just write something along the lines of "Found a cycle of natural landmark orderings." and nothing more. That the problem is unsolvable as a consequence will be printed anyway by the search algorithm, I think.

Reply:
I see, sure, makes sense.